### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by danieltichiyama**

### CODE REVIEW

**Review by edDevleague — CHANGES_REQUESTED** [review]
Commited node_modules...

**Comment by danieltichiyama** [comment]
Okay, `git rm -r --cached .` and re-commited.  Ready for review.

File: index.js [hunk 1]
```diff
@@ -0,0 +1 @@
+const database = require(knex);
```
Comment (edDevleague) [inline index.js]: This file isn't needed.

File: routes/carts.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
+const express = require("express");
+const router = express.Router();
+const db = require("../db");
+
+router.get("/carts/:user_id", (req, res) => {
+  db.raw("SELECT * FROM carts WHERE user_id = ?", [req.params.user_id]).then(
```
Comment (edDevleague) [inline routes/carts.js]: You should validate incoming user_id is a number.

File: routes/carts.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+router.post("/carts/:user_id/:product_id", (req, res) => {
+  let errMsg;
+  db.raw("SELECT * FROM users WHERE id = ?", [req.params.user_id])
+    .then(results => {
+      if (results.rowCount === 0) {
+        errMsg = "User not found";
+        throw new Error();
+      }
+      console.log("found user");
```
Comment (edDevleague) [inline routes/carts.js]: Remove console.log

File: routes/carts.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  db.raw("SELECT * FROM carts WHERE user_id = ?", [req.params.user_id]).then(
+    results => {
+      res.json(results.rows);
+    }
+  );
+});
+
+router.post("/carts/:user_id/:product_id", (req, res) => {
+  let errMsg;
+  db.raw("SELECT * FROM users WHERE id = ?", [req.params.user_id])
```
Comment (edDevleague) [inline routes/carts.js]: Validate your inputs.

File: routes/carts.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        throw new Error();
+      }
+      console.log("found user");
+      db.raw("SELECT * FROM products WHERE id = ?", [req.params.product_id])
+        .then(results => {
+          if (results.rowCount === 0) {
+            errMsg = "Product not found";
+            throw new Error();
+          }
+          console.log("found product");
```
Comment (edDevleague) [inline routes/carts.js]: Remove console.log

File: routes/carts.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+router.post("/carts/:user_id/:product_id", (req, res) => {
+  let errMsg;
+  db.raw("SELECT * FROM users WHERE id = ?", [req.params.user_id])
+    .then(results => {
+      if (results.rowCount === 0) {
+        errMsg = "User not found";
+        throw new Error();
+      }
+      console.log("found user");
+      db.raw("SELECT * FROM products WHERE id = ?", [req.params.product_id])
```
Comment (edDevleague) [inline routes/carts.js]: Validate your inputs.

File: routes/carts.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        .then(results => {
+          if (results.rowCount === 0) {
+            errMsg = "Product not found";
+            throw new Error();
+          }
+          console.log("found product");
+          db.raw("INSERT INTO carts (user_id, products_id) VALUES (?,?)", [
+            req.params.user_id,
+            req.params.product_id
```
Comment (edDevleague) [inline routes/carts.js]: Validate your inputs

File: routes/carts.js (hunk 6 above)
Comment (edDevleague) [inline routes/carts.js]: Return db.raw and you won't have to nest this.

File: routes/carts.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+      }
+      console.log("found user");
+      db.raw("SELECT * FROM products WHERE id = ?", [req.params.product_id])
+        .then(results => {
+          if (results.rowCount === 0) {
+            errMsg = "Product not found";
+            throw new Error();
+          }
+          console.log("found product");
+          db.raw("INSERT INTO carts (user_id, products_id) VALUES (?,?)", [
```
Comment (edDevleague) [inline routes/carts.js]: Return db.raw and you won't have to nest this.

File: routes/carts.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            req.params.product_id
+          ]).then(results => {
+            res.json({ success: true });
+          });
+        })
+        .catch(err => {
+          res.status(400).json({ message: errMsg });
+        });
+    })
+    .catch(err => {
```
Comment (edDevleague) [inline routes/carts.js:54]: You should only need a single catch

File: routes/carts.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+      res.status(400).json({ message: errMsg });
+    });
+});
+
+router.delete("/carts/:user_id/:product_id", (req, res) => {
+  let errMsg;
+  db.raw("SELECT * FROM users WHERE id = ?", [req.params.user_id])
```
Comment (edDevleague) [inline routes/carts.js]: Always return your promises, and validate  your inputs.

File: routes/products.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    .then(results => {
+      res.json(results.rows);
+    })
+    .catch(err => {
+      res.status(500).json({ message: "Something went wrong." });
+    });
+});
+
+router.get("/products/:product_id", (req, res) => {
+  db.raw("SELECT * FROM products WHERE id= ?", [req.params.product_id])
```
Comment (edDevleague) [inline routes/products.js]: Always return your promises, and validate  your inputs.

File: routes/products.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    })
+    .catch(err => {
+      res.status(400).json({ message: "Product not found" });
+    });
+});
+
+router.post("/products/new", (req, res) => {
+  db.raw(
+    "INSERT INTO products (title, description, inventory, price) VALUES (?,?,?,?) RETURNING *",
+    [req.body.title, req.body.description, req.body.inventory, req.body.price]
```
Comment (edDevleague) [inline routes/products.js]: Always return your promises, and validate  your inputs.

File: routes/products.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+router.put("/products/:product_id", (req, res) => {
+  db.raw(
+    "UPDATE products SET title = ?, description = ?, inventory = ?, price = ? WHERE id = ?",
+    [
+      req.body.title,
+      req.body.description,
+      req.body.inventory,
+      req.body.price,
+      req.params.product_id
```
Comment (edDevleague) [inline routes/products.js]: Always return your promises, and validate  your inputs.

File: routes/products.js [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+  });
+});
+
+router.delete("/products/:product_id", (req, res) => {
+  db.raw("SELECT * FROM products WHERE id = ?", [req.params.product_id])
+    .then(results => {
+      if (results.rowCount === 0) {
+        throw new Error();
+      }
+      db.raw("DELETE FROM products WHERE id = ?", [req.params.product_id]).then(
```
Comment (edDevleague) [inline routes/products.js]: Always return your promises, and validate  your inputs.

File: routes/users.js [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+const router = express.Router();
+const db = require("../db");
+
+router.get("/users/:user_id", (req, res) => {
+  db.raw("SELECT * FROM users WHERE id = ?", [req.params.user_id])
+    .then(results => {
+      if (results.rows.length === 0) {
+        throw new Error();
+      }
+      console.log(results.rows);
```
Comment (edDevleague) [inline routes/users.js]: Remove console.log

File: routes/users.js [hunk 16]
```diff
@@ -0,0 +1,90 @@
+const express = require("express");
+const router = express.Router();
+const db = require("../db");
+
+router.get("/users/:user_id", (req, res) => {
+  db.raw("SELECT * FROM users WHERE id = ?", [req.params.user_id])
```
Comment (edDevleague) [inline routes/users.js]: Always return your promises, and validate  your inputs.

File: routes/users.js [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+      res.json(results.rows);
+    })
+    .catch(err => {
+      console.log(err);
+      res.status(500).json({ message: "User not found" });
+    });
+});
+
+router.post("/users/login", (req, res) => {
+  console.log(req.body);
```
Comment (edDevleague) [inline routes/users.js]: console.log

File: routes/users.js [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+
+  let errMessage;
+  db.raw("SELECT * FROM users WHERE email = ?", [req.body.email])
```
Comment (edDevleague) [inline routes/users.js]: Always return your promises, and validate  your inputs.

File: routes/users.js [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+    .then(results => {
+      console.log(results.rows);
```
Comment (edDevleague) [inline routes/users.js]: console.log

File: routes/users.js [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+        errMessage = "User not found";
+        throw new Error();
+      }
+
+      if (results.rows[0].password !== req.body.password) {
+        errMessage = "Incorrect password";
+        throw new Error();
+      }
+
+      console.log(results.rows);
```
Comment (edDevleague) [inline routes/users.js]: console.log

File: routes/users.js [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+    });
+});
+
+router.post("/users/register", (req, res) => {
+  db.raw("SELECT * FROM users WHERE email = ?", [req.body.email])
+    .then(results => {
+      if (results.rows.length !== 0) {
+        throw new Error();
+      }
+      db.raw("INSERT INTO users (email,password) VALUES (?,?) RETURNING *", [
```
Comment (edDevleague) [inline routes/users.js]: Always return your promises, and validate  your inputs.

File: routes/users.js [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+    .catch(err => {
+      console.log(err);
+      res.status(400).json({ message: "User already exists" });
+    });
+});
+
+router.put("/users/:user_id/forgot-password", (req, res) => {
+  db.raw("UPDATE users SET password = ? WHERE id = ?", [
+    req.body.password,
+    req.params.user_id
```
Comment (edDevleague) [inline routes/users.js]: Always return your promises, and validate  your inputs.

File: routes/users.js [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+  ]).then(results => {
+    res.json({ message: "New password created!" });
+  });
+});
+
+router.delete("/users/:user_id", (req, res) => {
+  db.raw("DELETE FROM users WHERE id = ?", [req.params.user_id])
```
Comment (edDevleague) [inline routes/users.js]: Always return your promises, and validate  your inputs.

File: routes/users.js [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+    .then(results => {
+      console.log(results);
```
Comment (edDevleague) [inline routes/users.js]: console.log

File: server.js [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+
+const PORT = 3000;
+
+app.use(bodyParser.json());
+
+app.get("/", (req, res) => {
+  //   db.raw("SELECT * FROM users").then(results => {
+  // console.log(results.rows);
+  // res.json(results.rows);
+  //   });
```
Comment (edDevleague) [inline server.js]: Remove commented code

**Review by edDevleague — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (danieltichiyama) [reply index.js]: done.

File: routes/carts.js (hunk 3 above)
Comment (danieltichiyama) [reply routes/carts.js]: done.

File: routes/carts.js (hunk 5 above)
Comment (danieltichiyama) [reply routes/carts.js]: done.

File: server.js (hunk 25 above)
Comment (danieltichiyama) [reply server.js]: done.

**Review by edDevleague — CHANGES_REQUESTED** [review]
Still missing one console.log and you should be returning all your nested promises. This will keep your code from being indented and you can use a single catch to catch all your errors.

**Review by edDevleague — APPROVED** [review]

**Pull request merged**